File: src/c/conformance/scenario_engine.c
/********************
 *  Copyright 2026 TekFive, Inc., Sean M. Brennan, and contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *******************/

#include "scenario_engine.h"

#include <stdio.h>
#include <string.h>

void sce_init(sce_run_ctx_t *ctx) {
    memset(ctx, 0, sizeof(*ctx));
    for (int i = 0; i < SCE_MAX_STEPS; i++) {
        ctx->outbox_start[i] = -1;
        ctx->outbox_end[i] = -1;
    }
}

sce_participant_t *sce_find_participant(sce_run_ctx_t *ctx, const char *id) {
    if (ctx == NULL || id == NULL) return NULL;
    for (size_t i = 0; i < ctx->participant_count; i++) {
        if (strcmp(ctx->participants[i].id, id) == 0) {
            return &ctx->participants[i];
        }
    }
    return NULL;
}

void sce_capture(sce_run_ctx_t *ctx, const char *to_id, const char *function) {
    if (ctx == NULL) return;
    if (ctx->captured_count >= SCE_MAX_CAPTURED) return;
    sce_captured_t *cm = &ctx->captured[ctx->captured_count++];
    snprintf(cm->from, sizeof(cm->from), "%s", ctx->current_dispatcher);
    snprintf(cm->to, sizeof(cm->to), "%s", to_id ? to_id : "");
    snprintf(cm->function, sizeof(cm->function), "%s", function ? function : "");
}

/* ------------------------------------------------------------------------- */
/* Step drivers                                                               */
/* ------------------------------------------------------------------------- */

/* Build & dispatch an inbound to all matching targets. Records outbox bounds
 * for `step_id`. `repeat` controls how many times the same inbound is
 * dispatched — 1 for normal source steps, N for replay-idempotency probes.
 * The inbound is built once and dispatched N times (matches Python's
 * scenario_engine and a real network replay's "same wire bytes redelivered"
 * semantic). Outbox bounds wrap all N deliveries' captured emissions.
 * Returns 0 on success, -1 with ctx->err on dispatch failure. */
static int _build_and_deliver(sce_run_ctx_t *ctx, int step_id,
                              const char *from, const char *to,
                              const char *function, json_t *payload,
                              int repeat) {
    generic_msg_t inbound;
    if (ctx->build_inbound(ctx, from, to, function, payload, &inbound) != 0) {
        snprintf(ctx->err, sizeof(ctx->err),
                 "step %d: build_inbound(%s) failed", step_id, function);
        return -1;
    }

    ctx->outbox_start[step_id] = (int)ctx->captured_count;
    ctx->current_step_id = step_id;

    sce_participant_t *unicast_target = NULL;
    if (strcmp(to, "broadcast") != 0) {
        unicast_target = sce_find_participant(ctx, to);
        if (unicast_target == NULL) {
            snprintf(ctx->err, sizeof(ctx->err),
                     "step %d: unknown participant to:%s", step_id, to);
            return -1;
        }
    }

    for (int rep = 0; rep < repeat; rep++) {
        if (unicast_target == NULL) {
            /* Deliver to every participant except `from`. */
            for (size_t i = 0; i < ctx->participant_count; i++) {
                if (strcmp(ctx->participants[i].id, from) == 0) continue;
                snprintf(ctx->current_dispatcher, sizeof(ctx->current_dispatcher),
                         "%s", ctx->participants[i].id);
                ctx->dispatch(ctx, &ctx->participants[i], &inbound);
            }
        } else {
            snprintf(ctx->current_dispatcher, sizeof(ctx->current_dispatcher),
                     "%s", unicast_target->id);
            ctx->dispatch(ctx, unicast_target, &inbound);
        }
    }
    ctx->outbox_end[step_id] = (int)ctx->captured_count;
    return 0;
}

static int _drive_source(sce_run_ctx_t *ctx, json_t *step) {
    int sid = (int)json_integer_value(json_object_get(step, "id"));
    const char *from = json_string_value(json_object_get(step, "from"));
    const char *to = json_string_value(json_object_get(step, "to"));
    const char *function = json_string_value(json_object_get(step, "function"));
    json_t *payload = json_object_get(step, "payload");
    if (sid <= 0 || sid >= SCE_MAX_STEPS || from == NULL || to == NULL || function == NULL) {
        snprintf(ctx->err, sizeof(ctx->err), "source step missing required field");
        return -1;
    }
    int repeat = 1;
    json_t *rep_j = json_object_get(step, "repeat");
    if (json_is_integer(rep_j)) {
        int rv = (int)json_integer_value(rep_j);
        if (rv >= 1) repeat = rv;
    }
    return _build_and_deliver(ctx, sid, from, to, function, payload, repeat);
}

static int _drive_assertion(sce_run_ctx_t *ctx, json_t *step, int in_resp) {
    int sid = (int)json_integer_value(json_object_get(step, "id"));
    const char *from = json_string_value(json_object_get(step, "from"));
    const char *to = json_string_value(json_object_get(step, "to"));
    const char *function = json_string_value(json_object_get(step, "function"));
    if (sid <= 0 || sid >= SCE_MAX_STEPS || from == NULL || to == NULL || function == NULL) {
        snprintf(ctx->err, sizeof(ctx->err), "assertion step missing required field");
        return -1;
    }
    if (in_resp <= 0 || in_resp >= SCE_MAX_STEPS || ctx->outbox_start[in_resp] < 0) {
        snprintf(ctx->err, sizeof(ctx->err),
                 "step %d: in_response_to %d not yet run", sid, in_resp);
        return -1;
    }

    /* Match against captured emissions in the parent step's outbox. The
     * "broadcast" relaxation goes both ways: a scenario asserting "to: X"
     * also matches a captured "broadcast" (since broadcasts reach X), and
     * a scenario asserting "to: broadcast" matches any captured to_id. */
    bool match = false;
    for (int i = ctx->outbox_start[in_resp]; i < ctx->outbox_end[in_resp]; i++) {
        const sce_captured_t *cm = &ctx->captured[i];
        if (strcmp(cm->from, from) != 0) continue;
        if (strcmp(cm->function, function) != 0) continue;
        if (strcmp(to, "broadcast") == 0
            || strcmp(cm->to, to) == 0
            || strcmp(cm->to, "broadcast") == 0) {
            match = true;
            break;
        }
    }
    if (!match) {
        char captured[256] = {0};
        size_t off = 0;
        for (int i = ctx->outbox_start[in_resp];
             i < ctx->outbox_end[in_resp] && off < sizeof(captured) - 32; i++) {
            const sce_captured_t *cm = &ctx->captured[i];
            int n = snprintf(captured + off, sizeof(captured) - off,
                             "%s%s->%s:%s",
                             off == 0 ? "" : ", ",
                             cm->from, cm->to, cm->function);
            if (n < 0) break;
            /* snprintf returns the would-be length, so clamp rather than let
             * off run past the buffer and underflow the next size argument. */
            if ((size_t)n >= sizeof(captured) - off) {
                off = sizeof(captured) - 1;
                break;
            }
            off += (size_t)n;
        }
        /* Precisions bound each interpolation to its field width (ids are
         * SCE_ID_LEN, functions SCE_FN_LEN, the list is `captured` above) so
         * the whole message provably fits in ctx->err. */
        snprintf(ctx->err, sizeof(ctx->err),
                 "step %d: expected %.31s->%.31s:%.63s in response to step %d; "
                 "outbox of step %d held [%.255s]",
                 sid, from, to, function, in_resp, in_resp,
                 captured[0] ? captured : "(empty)");
        return -1;
    }

    /* Optional no_propagate: skip delivery, just mark this step as having
     * an empty outbox so future assertions can reference it. */
    json_t *no_prop = json_object_get(step, "no_propagate");
    if (json_is_true(no_prop)) {
        ctx->outbox_start[sid] = (int)ctx->captured_count;
        ctx->outbox_end[sid] = (int)ctx->captured_count;
        return 0;
    }

    /* Propagate: deliver this step's message to its `to`. The payload comes
     * from the step's own payload field (mirroring Python engine, where
     * the captured raw message is delivered; here we re-build via the
     * adapter callback for type-safety). `repeat` is meaningless on an
     * assertion step (the inbound message is whatever the source emitted,
     * once), so pin to 1. */
    return _build_and_deliver(ctx, sid, from, to, function,
                              json_object_get(step, "payload"), 1);
}

static int _check_expected_state(sce_run_ctx_t *ctx) {
    json_t *exp = json_object_get(ctx->case_data, "expected_state");
    if (!json_is_object(exp)) return 0;
    /* The engine treats expected_state as advisory unless a future hook
     * registers per-protocol checks. Keeping this stub matches Python's
     * engine, which delegates to `_check_expected_state` on the participant
     * impl when the impl exposes one. C has no equivalent yet — wire it
     * via an additional ops callback if a scenario needs it. */
    (void)ctx;
    return 0;
}

/* ------------------------------------------------------------------------- */
/* Public driver                                                              */
/* ------------------------------------------------------------------------- */

int sce_run(sce_run_ctx_t *ctx) {
    if (ctx == NULL || ctx->case_data == NULL
        || ctx->build_inbound == NULL || ctx->dispatch == NULL) {
        if (ctx != NULL)
            snprintf(ctx->err, sizeof(ctx->err), "ctx not initialized");
        return -1;
    }

    json_t *steps_j = json_object_get(ctx->case_data, "steps");
    if (!json_is_array(steps_j)) {
        snprintf(ctx->err, sizeof(ctx->err), "scenario: steps array missing");
        return -1;
    }

    size_t n = json_array_size(steps_j);
    for (size_t i = 0; i < n; i++) {
        json_t *step = json_array_get(steps_j, i);
        json_t *in_resp_j = json_object_get(step, "in_response_to");
        int rc;
        if (json_is_integer(in_resp_j)) {
            rc = _drive_assertion(ctx, step, (int)json_integer_value(in_resp_j));
        } else {
            rc = _drive_source(ctx, step);
        }
        if (rc != 0) return -1;
    }
    return _check_expected_state(ctx);
}

File: src/autonomous-trust/conformance/harness/python/scenario_engine.py
# ******************
#  Copyright 2026 Sean M. Brennan and contributors
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
# ******************

"""Universal scenario engine.

Owns the protocol-agnostic walk over scenario steps. Per-protocol adapters
plug in by providing a `Participant` factory and a `dispatch()` callback.

The engine treats every step as one of:

  - Source step (no `in_response_to`): the named `from` participant is the
    originator. The engine builds the message and delivers it to `to`.
  - Assertion step (with `in_response_to: N`): the engine looks up the outbox
    captured when step N was delivered, finds a message that matches this
    step's (from, to, function), asserts it exists, and then delivers that
    captured message to `to` so any cascade continues.

`to: broadcast` expands to every participant other than `from`.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable, Iterable

from ..common.scenario_loader import Case


@dataclass
class CapturedMessage:
    """An outbound message observed on a participant's network queue."""
    from_id: str
    to_id: str  # participant id, or 'broadcast'
    function: str
    payload: Any  # the original message.obj (kept loose for v1)
    raw: Any = None  # adapter-specific underlying object (e.g. a Message)


@dataclass
class StepOutbox:
    step_id: int
    captured: list[CapturedMessage] = field(default_factory=list)


@dataclass
class ParticipantHandle:
    """Per-participant state carried by the engine.

    `dispatch` is the adapter-supplied callback that delivers one inbound
    message to this participant and returns the messages it emitted.
    """
    id: str
    role: str
    impl: Any  # whatever the adapter wants (an IdentityProcess, a mock, ...)
    dispatch: Callable[[Any], list[CapturedMessage]]


@dataclass
class ScenarioContext:
    case: Case
    participants: dict[str, ParticipantHandle]
    outboxes: dict[int, StepOutbox] = field(default_factory=dict)
    # adapter-supplied: builds a Message-like inbound from (from_id, to_id,
    # function, payload) for a source step
    build_inbound: Callable[..., Any] = None
    # adapter-supplied: extracts (from_id, to_id, function, payload) from a
    # CapturedMessage's `raw` so the engine can match assertion steps
    describe: Callable[[CapturedMessage], tuple[str, str, str, Any]] = None


def run_scenario(ctx: ScenarioContext) -> None:
    """Walk the scenario's steps, applying the source/assertion model above.

    Raises AssertionError on the first failure, with a step-id-tagged message.
    """
    steps = ctx.case.data['steps']
    for step in steps:
        sid = step['id']
        in_resp = step.get('in_response_to')
        ctx.outboxes.setdefault(sid, StepOutbox(step_id=sid))

        if in_resp is None:
            _drive_source(ctx, step)
        else:
            _drive_assertion(ctx, step, in_resp)

    _check_expected_state(ctx)


def _drive_source(ctx: ScenarioContext, step: dict) -> None:
    sid = step['id']
    sender = ctx.participants.get(step['from'])
    if sender is None:
        raise AssertionError(f'step {sid}: unknown sender {step["from"]!r}')
    inbound = ctx.build_inbound(
        from_id=step['from'],
        to_id=step['to'],
        function=step['function'],
        payload=step.get('payload', {}),
    )
    targets = _resolve_targets(ctx, step, exclude={step['from']})
    # repeat: N delivers the same inbound message N times to the same
    # targets. Used by replay-handling probes; a correctly-deduping
    # handler produces the same observable state after N deliveries as
    # after 1. Built once, dispatched N times — matches what a real
    # network replay attack would look like (same wire bytes redelivered).
    repeat = int(step.get('repeat', 1))
    for _ in range(repeat):
        for target in targets:
            emitted = target.dispatch(inbound)
            ctx.outboxes[sid].captured.extend(_tag_emitter(emitted, target.id))


def _drive_assertion(ctx: ScenarioContext, step: dict, in_resp: int) -> None:
    sid = step['id']
    parent = ctx.outboxes.get(in_resp)
    if parent is None:
        raise AssertionError(
            f'step {sid}: in_response_to references step {in_resp} that has not run yet'
        )
    expected_from = step['from']
    expected_to = step['to']
    expected_fn = step['function']

    matches = [
        cm for cm in parent.captured
        if cm.from_id == expected_from
        and cm.function == expected_fn
        and (expected_to == 'broadcast' or cm.to_id in (expected_to, 'broadcast'))
    ]
    if not matches:
        captured_summary = ', '.join(
            f'{cm.from_id}->{cm.to_id}:{cm.function}' for cm in parent.captured
        ) or '(empty)'
        raise AssertionError(
            f'step {sid}: expected {expected_from}->{expected_to}:{expected_fn} '
            f'in response to step {in_resp}; outbox of step {in_resp} held [{captured_summary}]'
        )

    if step.get('no_propagate'):
        return  # assertion-only step; emission verified, delivery skipped

    cm = matches[0]
    targets = _resolve_targets(ctx, step, exclude={expected_from})
    for target in targets:
        emitted = target.dispatch(cm.raw)
        ctx.outboxes[sid].captured.extend(_tag_emitter(emitted, target.id))


def _resolve_targets(ctx: ScenarioContext, step: dict, exclude: set[str]) -> list[ParticipantHandle]:
    to = step['to']
    if to == 'broadcast':
        return [p for pid, p in ctx.participants.items() if pid not in exclude]
    target = ctx.participants.get(to)
    if target is None:
        raise AssertionError(f'step {step["id"]}: unknown target {to!r}')
    return [target]


def _tag_emitter(captured: Iterable[CapturedMessage], emitter_id: str) -> list[CapturedMessage]:
    out = []
    for cm in captured:
        if not cm.from_id:
            cm = CapturedMessage(
                from_id=emitter_id, to_id=cm.to_id, function=cm.function,
                payload=cm.payload, raw=cm.raw,
            )
        out.append(cm)
    return out


def _check_expected_state(ctx: ScenarioContext) -> None:
    expected = ctx.case.data.get('expected_state')
    if not expected:
        return
    for participant_id, asserts in expected.items():
        if participant_id == 'group':
            # group state is checked by the adapter via a sentinel; v1 logs
            # but does not enforce. Adapter can post-validate.
            continue
        p = ctx.participants.get(participant_id)
        if p is None:
            raise AssertionError(f'expected_state references unknown participant {participant_id!r}')
        if hasattr(p.impl, '_check_expected_state'):
            p.impl._check_expected_state(asserts)
